### ISSUE TITLE
SSL_CERT_DIR can cause LibGit2 initialization error

### ISSUE THREAD

**Issue by omus** [JuliaLang/julia#20439]
It appears that broken symlinks within the SSL_CERT_DIR can cause Julia to raise an exception during startup:

```bash
$ mkdir /tmp/certs
$ cd /tmp/certs
$ ln -s bar foo
$ ln -s /doesnotexist bar
$ SSL_CERT_DIR=/tmp/certs julia  # Julia 0.6
fatal: error thrown and no exception handler available.
Base.InitError(mod=:LibGit2, error=ErrorException("error initializing LibGit2 module"))
rec_backtrace at /home/centos/buildbot/slave/package_tarball64/build/src/stackwalk.c:84
record_backtrace at /home/centos/buildbot/slave/package_tarball64/build/src/task.c:239
jl_throw at /home/centos/buildbot/slave/package_tarball64/build/src/task.c:565
__init__ at ./libgit2/libgit2.jl:669
unknown function (ip: 0x7f0b84fc7a88)
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:2217
jl_apply at /home/centos/buildbot/slave/package_tarball64/build/src/julia.h:1411 [inlined]
jl_module_run_initializer at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:85
_julia_init at /home/centos/buildbot/slave/package_tarball64/build/src/init.c:704
julia_init at /home/centos/buildbot/slave/package_tarball64/build/src/task.c:292
unknown function (ip: 0x401656)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x40170c)
```

Note that the paths in the exception above appear to be from the official build server as I encountered this in a Debian environment.

versioninfo:
```
Julia Version 0.6.0-dev.2499
Commit cdca71b (2017-02-01 15:33 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
```

**Comment by omus** [comment]
Problem appears limited to Julia 0.6

**Comment by omus** [comment]
Also can occur with invalid certificate files:
```
$ mkdir /tmp/certs
$ cd /tmp/certs
$ touch empty.pem
$ SSL_CERT_DIR=/tmp/certs julia
fatal: error thrown and no exception handler available.
Base.InitError(mod=:LibGit2, error=ErrorException("error initializing LibGit2 module"))
rec_backtrace at /home/centos/buildbot/slave/package_tarball32/build/src/stackwalk.c:84
record_backtrace at /home/centos/buildbot/slave/package_tarball32/build/src/task.c:239
jl_throw at /home/centos/buildbot/slave/package_tarball32/build/src/task.c:565
__init__ at ./libgit2/libgit2.jl:791
unknown function (ip: 0xb2664d65)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball32/build/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball32/build/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball32/build/src/gf.c:2225
jl_apply at /home/centos/buildbot/slave/package_tarball32/build/src/julia.h:1410 [inlined]
jl_module_run_initializer at /home/centos/buildbot/slave/package_tarball32/build/src/toplevel.c:85
_julia_init at /home/centos/buildbot/slave/package_tarball32/build/src/init.c:704
julia_init at /home/centos/buildbot/slave/package_tarball32/build/src/task.c:292
unknown function (ip: 0x8049033)
__libc_start_main at /lib/i386-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x80490f0)
```

**Comment by kshyatt** [comment]
Is this still a problem?

**Comment by iamed2** [comment]
`0.6.0-pre.beta.445` still has this issue at least (tested on JuliaBox)

**Comment by StefanKarpinski** [comment]
Should this be a release blocker?

**Comment by omus** [comment]
With my personal work it's not been too hard to work around once I discovered what the problem was. I think it doesn't need to block 0.6 but should be fixed for 0.7 or 0.6.x

**Comment by tkelman** [comment]
Maybe worth bringing up and trying to get solved at https://github.com/libgit2/libgit2/pull/4173, the latest state of the mbedtls patch upstream (we should revisit and help get that merged eventually). Is it mbedtls that has issues with trying to read symlinks, or the code added to libgit2 should be more careful about what it reads?

**Comment by omus** [comment]
The code added should be more careful about what it reads. It will crash with empty files or broken symlinks. I expect any invalid SSL cert file in the directory will cause the same exception.

**Comment by omus** [comment]
Fixed in latest v0.7.
